File: android/java/org/chromium/chrome/browser/crypto_wallet/HnsWalletServiceFactory.java
/* Copyright (c) 2021 The Hns Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at https://mozilla.org/MPL/2.0/. */

package org.chromium.chrome.browser.crypto_wallet;

import org.chromium.base.annotations.JNINamespace;
import org.chromium.base.annotations.NativeMethods;
import org.chromium.hns_wallet.mojom.HnsWalletP3a;
import org.chromium.hns_wallet.mojom.HnsWalletService;
import org.chromium.chrome.browser.crypto_wallet.util.Utils;
import org.chromium.chrome.browser.profiles.Profile;
import org.chromium.mojo.bindings.ConnectionErrorHandler;
import org.chromium.mojo.bindings.Interface;
import org.chromium.mojo.bindings.Interface.Proxy.Handler;
import org.chromium.mojo.system.MessagePipeHandle;
import org.chromium.mojo.system.impl.CoreImpl;

@JNINamespace("chrome::android")
public class HnsWalletServiceFactory {
    private static final Object sLock = new Object();
    private static HnsWalletServiceFactory sInstance;

    public static HnsWalletServiceFactory getInstance() {
        synchronized (sLock) {
            if (sInstance == null) {
                sInstance = new HnsWalletServiceFactory();
            }
        }
        return sInstance;
    }

    private HnsWalletServiceFactory() {}

    public HnsWalletService getHnsWalletService(ConnectionErrorHandler connectionErrorHandler) {
        Profile profile = Utils.getProfile(false); // always use regular profile
        long nativeHandle =
                HnsWalletServiceFactoryJni.get().getInterfaceToHnsWalletService(profile);
        MessagePipeHandle handle = wrapNativeHandle(nativeHandle);
        HnsWalletService hnsWalletService = HnsWalletService.MANAGER.attachProxy(handle, 0);
        Handler handler = ((Interface.Proxy) hnsWalletService).getProxyHandler();
        handler.setErrorHandler(connectionErrorHandler);

        return hnsWalletService;
    }

    public HnsWalletP3a getHnsWalletP3A(ConnectionErrorHandler connectionErrorHandler) {
        Profile profile = Utils.getProfile(false); // always use regular profile
        long nativeHandle =
                HnsWalletServiceFactoryJni.get().getInterfaceToHnsWalletP3A(profile);
        MessagePipeHandle handle = wrapNativeHandle(nativeHandle);
        HnsWalletP3a hnsWalletP3A = HnsWalletP3a.MANAGER.attachProxy(handle, 0);
        Handler handler = ((Interface.Proxy) hnsWalletP3A).getProxyHandler();
        handler.setErrorHandler(connectionErrorHandler);

        return hnsWalletP3A;
    }

    private MessagePipeHandle wrapNativeHandle(long nativeHandle) {
        return CoreImpl.getInstance().acquireNativeHandle(nativeHandle).toMessagePipeHandle();
    }

    @NativeMethods
    interface Natives {
        long getInterfaceToHnsWalletService(Profile profile);
        long getInterfaceToHnsWalletP3A(Profile profile);
    }
}

File: browser/hns_wallet/hns_wallet_service_factory.cc
/* Copyright (c) 2022 The Hns Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at https://mozilla.org/MPL/2.0/. */

#include "hns/browser/hns_wallet/hns_wallet_service_factory.h"

#include <memory>
#include <utility>

#include "base/no_destructor.h"
#include "hns/browser/hns_wallet/hns_wallet_context_utils.h"
#include "hns/browser/hns_wallet/json_rpc_service_factory.h"
#include "hns/browser/hns_wallet/keyring_service_factory.h"
#include "hns/browser/hns_wallet/tx_service_factory.h"
#include "hns/components/hns_wallet/browser/hns_wallet_service.h"
#include "hns/components/hns_wallet/browser/hns_wallet_service_delegate.h"
#include "chrome/browser/browser_process.h"
#include "chrome/browser/profiles/incognito_helpers.h"
#include "components/keyed_service/content/browser_context_dependency_manager.h"
#include "components/user_prefs/user_prefs.h"
#include "content/public/browser/storage_partition.h"

namespace hns_wallet {

// static
HnsWalletServiceFactory* HnsWalletServiceFactory::GetInstance() {
  static base::NoDestructor<HnsWalletServiceFactory> instance;
  return instance.get();
}

// static
mojo::PendingRemote<mojom::HnsWalletService>
HnsWalletServiceFactory::GetForContext(content::BrowserContext* context) {
  if (!IsAllowedForContext(context)) {
    return mojo::PendingRemote<mojom::HnsWalletService>();
  }

  return static_cast<HnsWalletService*>(
             GetInstance()->GetServiceForBrowserContext(context, true))
      ->MakeRemote();
}

// static
HnsWalletService* HnsWalletServiceFactory::GetServiceForContext(
    content::BrowserContext* context) {
  if (!IsAllowedForContext(context)) {
    return nullptr;
  }
  return static_cast<HnsWalletService*>(
      GetInstance()->GetServiceForBrowserContext(context, true));
}

// static
void HnsWalletServiceFactory::BindForContext(
    content::BrowserContext* context,
    mojo::PendingReceiver<mojom::HnsWalletService> receiver) {
  auto* hns_wallet_service =
      HnsWalletServiceFactory::GetServiceForContext(context);
  if (hns_wallet_service) {
    hns_wallet_service->Bind(std::move(receiver));
  }
}

HnsWalletServiceFactory::HnsWalletServiceFactory()
    : BrowserContextKeyedServiceFactory(
          "HnsWalletService",
          BrowserContextDependencyManager::GetInstance()) {
  DependsOn(KeyringServiceFactory::GetInstance());
  DependsOn(JsonRpcServiceFactory::GetInstance());
  DependsOn(TxServiceFactory::GetInstance());
}

HnsWalletServiceFactory::~HnsWalletServiceFactory() = default;

KeyedService* HnsWalletServiceFactory::BuildServiceInstanceFor(
    content::BrowserContext* context) const {
  auto* default_storage_partition = context->GetDefaultStoragePartition();
  auto shared_url_loader_factory =
      default_storage_partition->GetURLLoaderFactoryForBrowserProcess();
  return new HnsWalletService(
      shared_url_loader_factory, HnsWalletServiceDelegate::Create(context),
      KeyringServiceFactory::GetServiceForContext(context),
      JsonRpcServiceFactory::GetServiceForContext(context),
      TxServiceFactory::GetServiceForContext(context),
      user_prefs::UserPrefs::Get(context), g_browser_process->local_state());
}

content::BrowserContext* HnsWalletServiceFactory::GetBrowserContextToUse(
    content::BrowserContext* context) const {
  return chrome::GetBrowserContextRedirectedInIncognito(context);
}

}  // namespace hns_wallet
